src/accountsdb/db.zig

```diff
@@ -0,0 +1,1645 @@
+                try thread_disk_dirs.append(thread_disk_dir);
+                disk_config = DiskMemoryConfig{ .dir_path = thread_disk_dir, .capacity = 0 };
+            }
+
+            // pre-alloc happens in the thread
+            const t_index = try LoadingThreadAccountsDB.init(
+                page_allocator,
+                n_bins,
+                .{},
+                disk_config,
+            );
+            thread_dbs.appendAssumeCapacity(t_index);
+        }
+        defer {
+            // memory is copied to the main thread allocator
+            // so we can deinit them here
+            for (thread_dbs.items) |*ti| {
+                ti.deinit();
+            }
+            thread_dbs.deinit();
+
+            if (use_disk_index) {
+                for (thread_disk_dirs.items) |thread_disk_dir| {
+                    std.fs.cwd().deleteTree(thread_disk_dir) catch |err| {
+                        self.logger.errf("failed to delete disk mem {s}: {s}\n", .{ thread_disk_dir, @errorName(err) });
+                    };
+                    self.allocator.free(thread_disk_dir);
+                }
+                thread_disk_dirs.deinit();
+            }
+        }
+
+        self.logger.infof("reading and indexing accounts...", .{});
+        var handles = try spawnThreadTasks(
+            self.allocator,
+            parseAndBinAccountFilesMultiThread,
+            .{
+                &fields,
+                accounts_path,
+                thread_dbs.items,
+                filenames.items,
+            },
+            filenames.items.len,
+            n_parse_threads,
+        );
+
+        for (handles.items) |handle| {
+            handle.join();
+        }
+        handles.deinit();
+        std.debug.print("\n", .{});
+        self.logger.infof("total time: {s}", .{std.fmt.fmtDuration(timer.read())});
+        timer.reset();
+
+        self.logger.infof("combining thread accounts...", .{});
+        try combineThreadDBs(self, thread_dbs.items, n_combine_threads);
+        std.debug.print("\n", .{});
+        self.logger.debugf("combining thread indexes took: {s}", .{std.fmt.fmtDuration(timer.read())});
+        timer.reset();
+    }
+
+    /// multithread entrypoint into parseAndBinAccountFiles
+    pub fn parseAndBinAccountFilesMultiThread(
+        fields: *const AccountsDbFields,
+        accounts_dir_path: []const u8,
+        thread_dbs: []LoadingThreadAccountsDB,
+        file_names: [][]const u8,
+        // task specific
+        start_index: usize,
+        end_index: usize,
+        thread_id: usize,
+    ) !void {
+        const thread_db = &thread_dbs[thread_id];
+        const thread_filenames = file_names[start_index..end_index];
+        try parseAndBinAccountFiles(
+            fields,
+            accounts_dir_path,
+            thread_db,
+            thread_filenames,
+            // NOTE: this constant has a large impact on performance due to allocations (best to overestimate)
+            1_500,
```

Comment:
yeah ive only checked mainnet to verify its large enough (its approx 1.5x what we need min)

i came up with this number just based on what i saw across mainnet snapshots - there is a max file size for account files (16GB) which might contribute to this number (dont think it gets anywhere close to that per file tho)

cli probs makes sense imo - dynamically calculating this number by counting all the accounts would be pretty expensive

Reply:
yeah thats a great point - i think this info isnt in the snapshot metadata because the rust code doesnt prealloc the way we do

this leads to an interesting idea of what additional information we could add to 'sig snapshots' to be even more efficient